src/codegen/compile.ml

```diff
@@ -686,21 +683,25 @@ module E = struct
     | [t] -> ValBlockType (Some t)
     | ts -> VarBlockType (nr (func_type env (FuncType ([], ts))))
 
-  let if_ env tys thn els = G.if_ (as_block_type env tys) thn els
 
-  (* NB: confuses wasm-opt, don't use for now
-  let _multi_if_ env tys1 tys2 thn els =
-    G.if_
-      (VarBlockType (nr (func_type env (FuncType (tys1, tys2)))))
-      thn els
-  *)
+  let prepare_branch_condition =
+    G.i (Convert (Wasm_exts.Values.I32 I32Op.WrapI64))
+  let if0 then_block else_block =
+    prepare_branch_condition ^^
+    G.if0 then_block else_block
+  let if1 return_type then_block else_block =
+    prepare_branch_condition ^^
+    G.if1 return_type then_block else_block
```

Comment:
I thing this only need because you widen the result of tests to 64-bit - maybe don't that (see compile_relop etc below)?

Reply:
Yes, this is related to using `i64` uniformly even for comparison results. It helped to get a general comparison that can also be used without a subsequent branch. Of course, the peephole optimization rule is important to later again eliminate the widening and narrowing.